tools: build and use setsid command from external source when not available

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PRs tries to fix #8424 by introducing some make logic to fetch, build and use setsid from an external repository. Apparently setsid is not available by default on MacOSX.

The Make logic in dist/tools and makefiles should be ok. I'm not sure whether the changes in Makefile.include could be improved.

Maybe the external repository is not the best one but this can be changed easily.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

1. Start a debug session on Linux (which provides setsid) using the `debug` target. The external setsid should not be used
2. Start a debug session on MacOSX (which doesn't setsid by default). The external setsid should be fetched and built on the fly.

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

fixes #8424

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

aabadie commented:
I wrote this PR blindly, since I don't have a Mac but maybe someone with a Mac can test ? @smlng or @kYc0o ?

cladmi commented:
I think the main problem is "do we even need `setsid`".
But making it work by default on `osx` while we figure that out is a good idea.
Fix the bug first!

There are also other places using `setsid` without using the `SETSID` variables.

cladmi commented on makefiles/tools/targets.inc.mk:
```diff
@@ -24,3 +24,8 @@ $(RIOTTOOLS)/mosquitto_rsmb/mosquitto_rsmb:
 mosquitto_rsmb: $(RIOTTOOLS)/mosquitto_rsmb/mosquitto_rsmb
 	@make -C $(RIOTTOOLS)/mosquitto_rsmb run
+
+$(RIOTTOOLS)/setsid/setsid:
```
Please also add the dependency to the `Makefile` like for `edbg` to make it rebuild on version change.

cladmi commented on Makefile.include:
```diff
@@ -466,6 +466,14 @@ define check_cmd
 
+# Check setsid command is available
+#   This is not the case on MacOSX, so it must be built on the fly
+ifeq (, $(shell which setsid))
```
The `shell which` patterns are a bit different, check the other ones. I think some system provides an error message on error so cannot rely on the output.
```suggestion
ifeq (0,$(shell which setsid 2>&1 > /dev/null ; echo $$?))
```

aabadie replied:
Thanks, but in fact the right line is:
```mk
ifneq (0,$(shell which setsid 2>&1 > /dev/null ; echo $$?))
```

aabadie commented:
@cladmi, thanks for your feedback. I've updated the PR with your suggestions and replaced remaining occurrences of setsid (with AVR based boards using avarice, which I think could be factorized but well).
It would be great if a Mac user could test this PR.

aabadie commented:
Maybe @x3ro can also test on Mac ?

aabadie commented:
@cladmi, do you think this PR is valid to go into the release ? (since it fixes a bug)
If yes, it would be great to have someone testing it.

x3ro commented:
I can confirm that this worked for me when testing with

     BOARD=stm32f429i-disc1 STLINK_VERSION=2 make debug

aabadie commented:
@cladmi, your comments are addressed and @x3ro tested it with success.

aabadie commented:
@cladmi I think your comments are addressed. Maybe @fjmolinas can also confirm that it works (since he works on a mac)

Lotterleben commented:
@aabadie Just tested on MAC OS, works for me :)

cladmi commented:
Can you squash ? for a final check of commit messages and all ?

I noticed 3 tools are using the same `debug.sh` file. Could be put in common somehow :D

aabadie commented:
> Can you squash ? for a final check of commit messages and all ?

finally done.

aabadie commented:
Another of my low hanging fruit I think. This was tested with success BTW. I can rebase if needed.

miri64 commented:
@smlng can you have a look?

aabadie commented:
This PR is waiting for an ACK for quite some time now.

aabadie commented:
Anyone could test and eventually merge this PR ?

aabadie commented on makefiles/tools/targets.inc.mk:
```diff
@@ -24,3 +24,8 @@ $(RIOTTOOLS)/mosquitto_rsmb/mosquitto_rsmb:
+$(RIOTTOOLS)/setsid/setsid: $(RIOTTOOLS)/setsid/Makefile
+	@echo "[INFO] setsid binary not found - building it from source now"
+	@make -C $(RIOTTOOLS)/setsid
+	@echo "[INFO] setsid binary successfully build!"
```
Note to myself: s/build/built/

aabadie replied:
Just checked and in fact, all other tools have the same typo. Let's leave it this way here and I'll open a follow-up PR to fix them all in a row.

fjmolinas replied:
yep

fjmolinas commented on Makefile.include:
```diff
@@ -507,6 +507,13 @@ define check_cmd
+# Check setsid command is available
+#   This is not the case on MacOSX, so it must be built on the fly
+ifneq (0,$(shell which setsid 2>&1 > /dev/null ; echo $$?))
+  export SETSID = $(RIOTTOOLS)/setsid/setsid
```
@cladmi is this the kind of places where target-export-variables should be used?

fjmolinas commented on Makefile.include:
```diff
@@ -507,6 +507,13 @@ define check_cmd
+#   This is not the case on MacOSX, so it must be built on the fly
+ifneq (0,$(shell which setsid 2>&1 > /dev/null ; echo $$?))
+  export SETSID = $(RIOTTOOLS)/setsid/setsid
+  DEBUGDEPS += $(SETSID)
```
This could be documented in `vars.inc.mk`

aabadie replied:
Documentation added.

fjmolinas commented on makefiles/tools/targets.inc.mk:
```diff
@@ -1,5 +1,5 @@
 # This file contains a collection of targets defined by tools shipped with RIOT.
-# The reason for keeping those targets separated from their tool Makfiles is
+# The reason for keeping those targets separated from their tool Makefiles is
```
unrelated

aabadie replied:
removed

fjmolinas commented on Makefile.include:
```diff
@@ -570,6 +570,15 @@ define check_cmd
+# This is not the case on MacOSX, so it must be built on the fly
+ifneq (,$(filter debug, $(MAKECMDGOALS)))
+  ifneq (0,$(shell which setsid 2>&1 > /dev/null ; echo $$?))
+    export SETSID = $(RIOTTOOLS)/setsid/setsid
```
I would still use `target-export-variables`. Now you are only making the check if `debug` is present in `$(MAKECMDGOALS)` which is good, but if `debug` is present as well as other targets, than it will still be evaluated and SETSID exported for all targets.

If you use `target-export-variables` this will only be exported for `debug`.

fjmolinas replied:
It will translate to:

`debug: export SETSID = $(RIOTTOOLS)/setsid/setsid`

aabadie replied:
using target-export-variables seems to work.

aabadie commented:
> Feel free to squash after addressing.

Done

aabadie commented:
Thanks again @fjmolinas !
